app/objects/beatmap.py

```diff
@@ -65,26 +65,38 @@ async def ensure_local_osu_file(
 ) -> bool:
     """Ensure we have the latest .osu file locally,
     downloading it from the osu!api if required."""
-    if (
-        not osu_file_path.exists()
-        or hashlib.md5(osu_file_path.read_bytes()).hexdigest() != bmap_md5
-    ):
-        # need to get the file from the osu!api
-        if app.settings.DEBUG:
-            log(f"Doing osu!api (.osu file) request {bmap_id}", Ansi.LMAGENTA)
-
-        url = f"https://old.ppy.sh/osu/{bmap_id}"
-        async with app.state.services.http_client.get(url) as resp:
-            if resp.status != 200:
-                if 400 <= resp.status < 500:
-                    # client error, report this to cmyui
-                    stacktrace = app.utils.get_appropriate_stacktrace()
-                    await app.state.services.log_strange_occurrence(stacktrace)
-                return False
-
-            osu_file_path.write_bytes(await resp.read())
-
-    return True
+
+    exists = osu_file_path.exists()
+    # some callers have type Any | str on "bmap_md5" so I assume it's optional, thus May be None.
+    # I use a magic string here to avoid letting optional md5 slip through.
+    file_md5 = (
+        hashlib.md5(osu_file_path.read_bytes()).hexdigest()
+        if exists
+        else "MAGIC_STRING__NOT_EXISTS"
+    )
+    if file_md5 == bmap_md5:
+        return True
+    # need to get the file from the osu!api
+    if app.settings.DEBUG:
+        log(f"Doing osu!api (.osu file) request {bmap_id}", Ansi.LMAGENTA)
+
+    url = f"https://old.ppy.sh/osu/{bmap_id}"
+    async with app.state.services.http_client.get(url) as resp:
+        if resp.headers.get("Content-Disposition", "").startswith("attachment"):
```

Comment:
old pr but curious -- what's this `Content-Disposition` case for? :eyes: is this an osu-specific behavior? or something else?

Reply:
With this header browser will save the file to disk with name if provided, instead of trying to show the content.

@NiceAesth told me I should check if the response is valid with this header.